src/com/xilinx/rapidwright/design/ModuleInst.java

```diff
@@ -389,6 +390,12 @@ public boolean place(Site newAnchorSite, boolean skipIncompatible, boolean allow
                 if (pipSet != null) {
                     pipSet.add(newPip);
                 }
+                // Some tiles have nodes that are depopulated, we need to detect those
+                Node endNode = newPip.getEndNode();
+                if (endNode != null && endNode.getAllDownhillPIPs().size() == 0
+                        && pip.getEndNode().getAllDownhillPIPs().size() != 0) {
```

Comment:
Is looking at the existence of downhill PIPs the best way to check for depopulation? It seems that the new PIP has its tile forcibly updated, but is there a more general way to check for the validity of this new PIP?

Reply:
I'm open to other suggestions, but I couldn't find another way.

Reply:
I can't see another way without adding a new method onto `Node` either, and it's not clear to me what should be checked when inside too. I have a nagging worry about performance here since every placement attempt (which `getAllValidPlacements` does) that gets through to routing will involve examining all downhill nodes of all PIPs at least once.

Reply:
I know performance is an issue, but it may not matter for this particular application.  `getAllValidPlacements()` is or can be called offline and results cached in the module.  Then, during placement, only the collisions between it and the logic already placed needs to be checked.  The block placer does a number of optimizations during placement to avoid ever calling the actual `place()` method because it is expensive.  For example, for a modestly small module, it takes ~3.2 secs to call `getAllValidPlacements()`, it calls `place()` 117120 times, 38400 being successful.  Before this PR, it took ~2.5 secs. 

I spent some time looking into ways to cache the architectural special cases, but I'm not sure the effort is warranted yet.